[1단계 - 상품 목록] 초코(강다빈) 미션 제출합니다.

00kang opened the pull request:
> 개인 리뷰: 짝이 되기 전에 먼저 자신의 코드를 살펴보고 개선할 부분을 미리 생각해 봅니다.
> 질문 목록 작성: 궁금한 점이나 명확히 이해되지 않는 부분을 질문 목록으로 만들어 봅니다. PR에 작성한 질문을 활용해도 좋습니다.

### 개선할 부분
- 무한스크롤을 구현하여 다음 페이지의 상품을 불러오긴 하는데, 불러올 때마다 리렌더링이 일어나서 인지(?) 스크롤이 초기화되어 맨 위로 올라가는 문제가 있어요. 

- 장바구니 API 연동부분은 시간 부족으로 아직 구현 중이에요.

- 추가로 개선할 부분을 짚어주면 좋겠어요~

### 궁금한 점
- 현재 sorting할 때마다 새로운 상품들이 불러와지는 것 같은데 이게 정상 동작하고 있는 건지 궁금해요.

- MSW를 사용하여 상품 목록 조회 API 요청에 대해 다양한 응답 시나리오(성공, 실패, 빈 데이터 등)에 대한 테스트 코드를 작성하는 게 목표였어요. 어떠한 시나리오가 올 수 있을까요?

- 현재 코드 기준으로 불필요한 리렌더링이 많이 일어나는 건 아닐까? 비동기 요청이 잘 처리되고 있는건가? 

- 내 코드지만 나도 모르겠을 때(,,ㅋㅋㅋ) 어떻게 정리를 해나가나요? 정리하는 팁이 있을까요?

00kang committed:
docs : 기능 요구 사항 작성

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
chore : styled components 설치

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : main.tsx에 reset.css 추가

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 사용되는 이미지 assets 폴더에 정리

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : BaseButton, CartButton, ShopButton 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : Header 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : dropdown 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : ProductHeader 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
refactor : BaseButton 컴포넌트 불필요한 props 제거

00kang committed:
feat : CartActionButton 컴포넌트 구현

00kang committed:
feat : CartButton 컴포넌트에 상품종류개수 ui 추가

00kang committed:
feat : ProductProps 타입 정의

00kang committed:
feat : ProductItem 컴포넌트 구현

00kang committed:
feat : ProductList 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : Product 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : ToastNotification 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
fix : Header 컴포넌트 위치 고정

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
style : App 컴포넌트 스타일 지정

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : vitest 설정

- handlers.ts : API 요청 처리 핸들러
- server.ts : msw 서버 객체
- vitest.config.ts : vitest 설정 파일
- vitest.setup.ts : msw 서버 관리

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : URL 상수처리

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 상품 목록 조회하는 useProducts 커스텀 훅 구현 및 테스트

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
refactor : ProductList 컴포넌트에서 useProducts 커스텀 훅 사용하여 상품 목록 조회

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
fix : mockData인 products.json 형식 수정

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : LoadingSpinner 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 상품 목록 조회 중 로딩 상태 useProducs 커스텀 훅에 추가

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 에러 메시지 정의 및 ToastNotification 컴포넌트 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : context API로 에러 상태를 관리하는 errorContext 구현

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 상품 목록 조회 중 에러 발생 상태 useProducts 커스텀 훅에 추가

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 로딩 상태와 에러 상태에 따른 Product 컴포넌트 조건부 렌더링

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
refactor : mockData 상품 100개로 수정

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
fix : 테스트 코드에 ErrorProvider 제공하지 않아 발생하던 문제 해결

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 페이지네이션 기능을 useProducts 커스텀 훅에 추가

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
refactor : import문 정렬 정리

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 무한 스크롤 페이지네이션

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
fix : API 경로 수정

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : 카테고리와 가격순 정렬 기능 구현

- Product 컴포넌트에서 ProductHeader와 ProductList로 props 전달

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : API 주소 설정

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
refactor : CartActionButton 컴포넌트 수정

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : CartButton 컴포넌트에 카운트 보여주기

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

00kang committed:
feat : CartActionButton 컴포넌트에 api 통신 연결

Co-Authored-By: soosoo22 <80167893+soosoo22@users.noreply.github.com>

llqqssttyy commented on src/hooks/useProducts.ts:
```diff
@@ -0,0 +1,90 @@
+      const data = await response.json();
+
+      setProducts((prevProducts) =>
+        page === 1 ? data.content : [...prevProducts, ...data.content]
```
서버 내부적으로 시작 페이지가 0이어서, 첫 데이터부터 가져오려면 프론트의 시작 페이지도 0으로 요청해주셔야 해요...ㅠㅜ

llqqssttyy commented on src/hooks/useProducts.ts:
```diff
@@ -0,0 +1,90 @@
+      const categoryQuery = category === "전체" ? "" : `category=${category}`;
+
+      const response = await fetch(
+        `${PRODUCTS_ENDPOINT}?${categoryQuery}&page=${page}&size=${size}&sort=${sortOption}`
```
(취사 선택)queryString을 만들어 주는 유틸 함수를 만들어 보면 어떨까요?

00kang committed:
style : ProductItem border 스타일 적용

00kang committed:
refactor : 적절한 태그 사용하도록 수정

- <select>에 aria-label 추가
- <img>에 alt="" 추가
- 상품 제목에는 <h2> 추가
- z-dinex  값 상수로 관리
- <section> 사용하는 컴포넌트 ProductSection 으로 이름 수정

00kang committed:
refactor : png 파일 svg 파일로 수정

00kang committed:
refactor : index.ts 활용해서 임포트문 정리

00kang committed:
feat : 상품 가격 포맷팅하는 formatPrice 사용

00kang committed:
refactor : mockData 실제 응답 스키마와 동일하게 수정

00kang committed:
refactor : sortOption type 구체적으로 수정

00kang committed:
fix : 상품 로드 개수 문제 해결(첫페이지는 20개의 상품을, 다음페이지부터는 4개의 상품을 불러오도록)

00kang committed:
fix : isLoading으로 인한 리렌더링때문에 무한스크롤이 작동하지 않는 문제 해결

00kang committed:
refactor : 변수 index가 좌항에 오도록 수정

00kang committed:
refactor : 장바구니 관련 로직 미해결!!

- 장바구니 수량이 아니라 장바구니 리스트를 기반으로
- 장바구니에 담겨있다면 빼기 버튼 활성화
- 장바구니에 없다면 담기 버튼 활성화
- 각 버튼에 맞는 api 요청

인데 400 에러 미해결 상태!

00kang committed:
docs : README.md 업데이트

00kang committed:
refactor : 주석 삭제

00kang committed:
fix : 장바구니 담고 빼는 버튼 상태 불안정한 상태 해결

- 장바구니 상품 동기화 로직 순서 수정
- 버튼 동작 코드 분리

00kang committed:
fix : context API 활용해서 장바구니 상품 수량 카운트하도록 수정

00kang committed:
refactor : 의존성 배열에 isInCart 추가

00kang committed:
refactor : error 타입 추가
